File: toupper/src/lib/drinfo/instruction.ts
import type { Brush } from "./brush";
import type { Point } from "./point";

export type Stroke = {
  points: Point[];
  brush: Brush;
};

export type Motion = {
  start: Point;
  end: Point;
  selection: [Point, Point];
};

export type ImageInsertion = {
  base64: string;
  point: Point;
  scale: Point;
  rotate: number;
};

export type Bucket = {
  point: Point;
  brush: Brush;
};

export type Instruction = Stroke | Motion | ImageInsertion | Bucket;

export type InstructionBox = {
  instruction: Instruction;
  uuid: string;
  applied: boolean;
};

File: drinfo/src/instructions/instruction.rs
use core::fmt::Debug;

use serde::{Deserialize, Serialize};

use crate::{Bucket, ImageInsertion, Motion, Stroke};

/// An instruction.
#[derive(Debug, Serialize, Deserialize, Clone)]
pub enum Instruction {
    Bucket(Bucket),
    ImageInsertion(ImageInsertion),
    Motion(Motion),
    Stroke(Stroke),
}

/// An instruction box.
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct InstructionBox {
    pub instruction: Instruction,
    pub applied: bool,
    pub uuid: String,
}
